azure-pipelines-templates/device-test.yml

```diff
@@ -0,0 +1,156 @@
+              sed -i "s|<RealHardwarePort>.*<\/RealHardwarePort>|<RealHardwarePort>$(echo $SerialPort)<\/RealHardwarePort>|g" $(System.DefaultWorkingDirectory)/Artifacts/${{ split(tests, ',')[0]}}
+              cat $(System.DefaultWorkingDirectory)/Artifacts/${{ split(tests, ',')[0]}}
+              exit $?
+            displayName: 'Update nano.runsettings'
+
+          - bash: |
+              # Maximum number of retries
+              MAX_RETRIES=${{ parameters.MaxRetries }}
+              # Delay between retries in seconds
+              DELAY=2
+              # Initialize the counter
+              attempt=0
+              # Define the base directory to search in
+              BASE_DIR=$(System.DefaultWorkingDirectory)/Artifacts
+
+              # Use the find command to search for paths starting with packages/nanoFramework.TestFramework
+              TESTADAPTER_PATH=$(find "$BASE_DIR" -type d -path "*/packages/nanoFramework.TestFramework*.*" | head -n 1)
+
+              # Extract the directory path from the unitTestRunsettings parameter
+              UNIT_TEST_RUNSETTINGS_DIR=$(dirname "$BASE_DIR/${{ split(tests, ',')[0] }}")
+
+              # Loop to retry the function call
+              while (( attempt < MAX_RETRIES )); do
+                  RESULT_TEST=$(mono /azp/TestPlatform/vstest.console.exe $BASE_DIR/${{ split(tests, ',')[1] }} /Settings:$BASE_DIR/${{ split(tests, ',')[0]}} /TestAdapterPath:$TESTADAPTER_PATH /Logger:trx)
+                  echo "$RESULT_TEST"
+                  if [[ "$RESULT_TEST" == *"Couldn't connect to the device"* || "$RESULT_TEST" == *"Couldn't find any valid nanoDevice"* ]]; then
+                      echo "Test failed. Retrying in $DELAY seconds..."
+                      # we need to delete the trx file to avoid confusion but we keep the last one
+                      if [ $attempt -eq $MAX_RETRIES ]; then
+                        echo "Test failed after $MAX_RETRIES attempts."
+                        exit 1
+                      else
+                        find "$UNIT_TEST_RUNSETTINGS_DIR"/TestResults -type f -name "*.trx" -exec rm -f {} \;
+                        ((attempt++))
+                        sleep $DELAY
+                      fi
+                  elif [[ "$RESULT_TEST" =~ "The argument .* is invalid." ]]; then
+                      echo "Test failed as arguments are not valid."
+                      exit 1
+                  else
+                      break
+                  fi
+              done                     
+            displayName: 'Run tests'
+            timeoutInMinutes: 120
+            retryCountOnTaskFailure: 3
+          
+          - task: PublishTestResults@2
+            inputs:
+              testResultsFormat: 'VSTest' # 'JUnit' | 'NUnit' | 'VSTest' | 'XUnit' | 'CTest'. Alias: testRunner. Required. Test result format. Default: JUnit.
```

Comment:
Can't we get rid of the commented out ones? Don't seem usefull and I guess we're "stuck" with VSTest anyway...